### PR TITLE
feat(TPG>=6.11)!: add endpoint_dns

### PR DESCRIPTION
**Pull request by apeabody**
Fixes: https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/2182

### CODE REVIEW

**Comment by apeabody** [comment]
Error doesn't appear related to this change:
```
Step #51 - "teardown beta-cluster":         	Error:      	Received unexpected error:
Step #51 - "teardown beta-cluster":         	            	FatalError{Underlying: error while running command: exit status 1; 
Step #51 - "teardown beta-cluster":         	            	Error: Error when reading or editing Subnetwork: googleapi: Error 400: The subnetwork resource 'projects/ci-gke-5db7e0f2-bhp6/regions/us-central1/subnetworks/cft-gke-test-ivye' is already being used by 'projects/ci-gke-5db7e0f2-bhp6/regions/us-central1/forwardingRules/aec5bcafcba7344aa801978cb45df853', resourceInUseByAnotherResource
Step #51 - "teardown beta-cluster":         	            	}
Step #51 - "teardown beta-cluster":         	Test:       	TestBetaCluster
```

**Comment by zcrisler** [comment]
Resolves #2182

**Comment by apeabody** [comment]
Starting with TPG v6.11.0+ this has been a reliable error during beta-cluster teardown:
```
Error: Error when reading or editing Subnetwork: googleapi: Error 400: The subnetwork resource 'projects/ci-gke-76cf7158-g9i9/regions/us-central1/subnetworks/cft-gke-test-wner' is already being used by 'projects/ci-gke-76cf7158-g9i9/regions/us-central1/forwardingRules/ac9b02a8710734e47a5f8449ff61e2ec', resourceInUseByAnotherResource
```

File: autogen/main/main.tf.tmpl [hunk 1]
```diff
@@ -144,7 +144,7 @@ locals {
   cluster_output_zones          = local.cluster_output_regional_zones
 
 {% if private_cluster %}
-  cluster_endpoint           = (var.enable_private_nodes && length(google_container_cluster.primary.private_cluster_config) > 0) ? (var.deploy_using_private_endpoint ? google_container_cluster.primary.private_cluster_config[0].private_endpoint : google_container_cluster.primary.private_cluster_config[0].public_endpoint) : google_container_cluster.primary.endpoint
+  cluster_endpoint           = var.deploy_using_private_endpoint || var.enable_private_endpoint ? google_container_cluster.primary.control_plane_endpoints_config[0].dns_endpoint_config[0].endpoint : google_container_cluster.primary.endpoint
```
[the remark this answers is not in the document]
Comment (gorge511) [reply autogen/main/main.tf.tmpl]: Thanks for incorporating my suggestion.

File: autogen/main/outputs.tf.tmpl [hunk 2]
```diff
@@ -76,6 +76,24 @@ output "endpoint" {
   ]
 }
 
+output "endpoint_dns" {
+  sensitive   = true
```
Comment (gorge511) [inline autogen/main/outputs.tf.tmpl:80]: @apeabody I have a question about the sensitivity. I know the `endpoint` output is also marked as sensitive, but do you think it makes sense for both outputs? It's marked as sensitive from the initial commit to this module without any description. I know the standard way of configuring kubectl is via `gcloud container clusters get-credentials`, but should the endpoint IP/DNS be really sensitive? 🤔

Comment (apeabody) [reply autogen/main/outputs.tf.tmpl:80]: Hi @gorge511 - I carried it over, but had the same through it wasn't really required (it's not a token, cert, password, etc).  I'm considering to remove from both.

Comment (apeabody) [reply autogen/main/outputs.tf.tmpl]: removed from endpoint_dns for now

**Comment by apeabody** [comment]
All pre-requisite PRs have been merged, this is now ready for review.

**Comment by allthatjazzleo** [comment]
how's it now?
can't wait to use this feature :)

**Review by bharathkkb — APPROVED** [review]

File: modules/beta-private-cluster/cluster.tf [hunk 3]
```diff
@@ -546,6 +546,15 @@ resource "google_container_cluster" "primary" {
     }
   }
 
+  dynamic "control_plane_endpoints_config" {
+    for_each = var.enable_private_endpoint && var.deploy_using_private_endpoint ? [1] : [0]
+    content {
+      dns_endpoint_config {
+        allow_external_traffic = var.deploy_using_private_endpoint
```
Comment (legal90) [inline modules/beta-private-cluster/cluster.tf:553]: @apeabody  Why is this feature gated by `var.enable_private_endpoint && var.deploy_using_private_endpoint` ?

It should de possible to have "DNS-based endpoint" feature enabled without enforcing the usage of the private IP / private endpoint 🤔

Comment (apeabody) [reply modules/beta-private-cluster/cluster.tf:553]: Hi @legal90 - I was planning to update this behavior as part of a larger change when https://github.com/hashicorp/terraform-provider-google/issues/20369 is released, but I'd be happy to review a PR for just this aspect mow.

Comment (legal90) [reply modules/beta-private-cluster/cluster.tf:553]: Hi @apeabody ,
Thank you for your reply! I found your PR, which fixes that behavior and makes it working as it really should (at least, to my view): https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/2313
Kudos! I don't have anything to add there.

P.s. I'm sorry if my previous message sounded unpleasant. Thank you for you hard work on this module! ❤️

**Pull request merged**